File: binary-search-trees/bst.js
class Node {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
}

class BST {
  constructor() {
    this.root = null;
  }
  insert(value) {
    var newLeaf = new Node(value);
    if (this.root === null) {
      this.root = newLeaf;
      return this;
    } else {
      var temp = this.root;
      while (true) {
        if (value === temp.value) return undefined;
        if (value < temp.value) {
          if (temp.left === null) {
            temp.left = newLeaf;
            return this;
          } else {
            temp = temp.left;
          }
        } else {
          if (temp.right === null) {
            temp.right = newLeaf;
            return this;
          } else {
            temp = temp.right;
          }
        }
      }
    }
    return this;
  }
  find(value) {
    var current = this.root;
    var found = false;
    if (this.root === null) return null;
    while (current && !found) {
      if (value < current.value) {
        current = current.left;
      } else if (value > current.value) {
        current = current.right;
      } else {
        found = true;
      }
    }
    return found;
  }
  breathFirstSearch() {
    var node = this.root;
    var data = [];
    var queue = [];
    queue.push(node);
    while (queue.length) {
      node = queue.shift();
      data.push(node.value);
      if (node.left) queue.push(node.left);
      if (node.right) queue.push(node.right);
    }
    console.log("done");
    return data;
  }
  dfsPreorder() {
    var visited = [];
    var current = this.root;
     function preorderHelper(node) {
      visited.push(node.value);
       if (node.left) preorderHelper(node.left);
       if (node.right) preorderHelper(node.right);
    }
    preorderHelper(current);
    return visited;
  }
}

var tree = new BST();
console.log(tree.insert(10));
console.log(tree.insert(9));
console.log(tree.insert(11));
console.log(tree.insert(33));
console.log(tree.dfsPreorder());

File: binary-search-trees/bst.go
package main

import (
	"fmt"
)

type Tree struct {
	node *Node
}

func (t *Tree) insert(value int) *Tree {
	if t.node == nil {
		t.node = &Node{value: value}
	} else {
		t.node.insert(value)
	}
	return t
}

type Node struct {
	value int
	left  *Node
	right *Node
}

func (n *Node) insert(value int) {
	if value <= n.value {
		if n.left == nil {
			n.left = &Node{value: value}
		} else {
			n.left.insert(value)
		}
	} else {
		if n.right == nil {
			n.right = &Node{value: value}
		} else {
			n.right.insert(value)
		}
	}
}

func (n *Node) exists(value int) bool {
	if n == nil {
		return false
	}
	if n.value == value {
		return true
	}
	if value <= n.value {
		return n.left.exists(value)
	} else {
		return n.right.exists(value)
	}
}

func printNode(n *Node) {
	if n == nil {
		return
	}
	fmt.Println(n.value)
	printNode(n.left)
	printNode(n.right)
}

func bfs(root *Node) []int {
	values := make([]int, 0, 0)
	queue := []*Node{}
	r := root
	queue = append(queue, r)
	for len(queue) != 0 {
		r = queue[0]
		queue = queue[1:]
		values = append(values, r.value)
		if r.left != nil {
			queue = append(queue, r.left)
		}
		if r.right != nil {
			queue = append(queue, r.right)
		}
	}
	return values
}

func preorderTraversal(root *Node) {
	if root == nil {
		return
	}
	fmt.Println(root.value)
	preorderTraversal(root.left)
	preorderTraversal(root.right)
}

func main() {
	t := &Tree{}
	t.insert(10).
		insert(5).
		insert(20).
		insert(2).
		insert(40).
		insert(1).
		insert(60)
	preorderTraversal(t.node)
}
